server/py/framework/utils/clients/discovery.py

```diff
@@ -0,0 +1,150 @@
+        self.register_service(service_name="api-chief")
+
+    def register_service(
+        self,
+        service_name: str,
+    ):
+        """Register a new service instance."""
+        method_routes = self._resolve_service_method_routes(service_name)
+        url = self._resolve_service_url(service_name)
+        self.services[service_name] = ServiceInstance(
+            name=service_name,
+            url=url,
+            method_routes=method_routes,
+        )
+        self._logger.info("Registered service", service_name=service_name, url=url)
+
+    def deregister_service(self, service_name: str):
+        """Deregister a service instance."""
+        if service_name in self.services:
+            del self.services[service_name]
+
+        self._logger.info(
+            "Deregistered service",
+            service_name=service_name,
+        )
+
+    def resolve_service_by_request(
+        self, method: str, path: str
+    ) -> typing.Optional[ServiceInstance]:
+        """
+        Resolve path and returns the matching service instance for the request.
+
+        :param method: HTTP method of the request
+        :param path: URL path to match against service patterns
+        :return: ServiceInstance matching the request or None if no match
+        """
+        method = method.lower()
+        if service_name := self._find_service(method, path):
+            return self.get_service(service_name)
+        return None
+
+    def get_service(self, service_name: str) -> typing.Optional[ServiceInstance]:
+        """Get all healthy instances of a service."""
+        return self.services.get(service_name, None)
+
+    def _resolve_service_method_routes(
+        self, service_name: str
+    ) -> dict[str, list[re.Pattern]]:
+        """Resolve service routes per method for a service"""
+        method_routes = {
+            "get": [],
+            "post": [],
+            "put": [],
+            "delete": [],
+        }
+        routes = self._service_routes(service_name)
+        for methods, path in routes:
+            if methods == ["*"]:
+                for method in method_routes:
+                    method_routes[method].append(re.compile(path))
+                continue
+
+            for method in methods:
+                method_routes[method].append(re.compile(path))
+
+        return method_routes
+
+    def _find_service(self, method: str, path: str):
+        """
+        Find first service matching the given request URL
+
+        :param method: HTTP method of the request
+        :param path: URL path to match against service patterns
+        :return: Name of matching service or None if no match
+        """
+        for service_name, service_instance in self.services.items():
+            routes_patterns = service_instance.method_routes[method]
+            for route_pattern in routes_patterns:
+                if route_pattern.match(path):
+                    return service_name
```

Comment:
We will call this function very often, and the values on it will not change during the lifecycle of the discovery service, consider adding caching to this function. (Invalidate the cache when registering or derigstering service)